scripts/train_lm.py

```diff
@@ -146,12 +146,39 @@ def WriteMetaparameters(metaparameters, ngram_order, num_train_sets, out_file):
 
 def FormatMetaparameters(metaparameters):
     out = []
+    # get the maximum length of decimals of metaparameters
+    decimal_length = 0
     for param in metaparameters:
-        x = '{:.3f}'.format(param)
-        if x == '0.00':
-            x = '{:.3g}'.format(param)
-        out.append(x)
-
+        temp_length = len(str(param))
+        if decimal_length < temp_length:
+            decimal_length = temp_length
+    decimal_length -= 2
+    
+    temp_original = metaparameters[0]
+    temp = round(temp_original, 3)
+    for param in metaparameters[1:]:
+        round_index = 3
+        x_original = param
+        x = round(x_original, round_index)
+        while(temp == x and round_index <= decimal_length):
```

Comment:
Hello, Dan and Jian
I think I came up with a ‘decent' solution to this which can handle some extreme cases I tested (see the tested cases after code),
Please let me if there is any issue or anything can be changed.

def FormatMetaparameters(metaparameters):
    # First, we check if there is any collison in metaparameters
    all_metaparameters = set([])
    collision_idx = []
    for idx in range(len(metaparameters)):
        if metaparameters[idx] in all_metaparameters:
            collision_idx.append(idx)
        else:
            all_metaparameters.add( metaparameters[idx] )
    # If collision exist, we want to record its idx in metaparameters, and prompt user
    if collision_idx: 
        LogMessage("Some metaparameters may be same, which can be suspicous")
    if 0.0 or 1.0 in all_metaparameters:
        LogMessage("Some metaparameters is 0.0 or 1.0, where is legitimate range is 0 < x < 1")
    decimal.getcontext().rounding = decimal.ROUND_DOWN

    out = []
    for idx in range(len(metaparameters)):
        # set initial precision be 3 and increase by 1 if we find new collision 
        # (for old collision we leave it as it is)
        num_significant_digit = 3
        decimal.getcontext().prec = num_significant_digit
        x = str(decimal.Decimal( metaparameters[idx] ) + decimal.Decimal(0) )
        while( idx not in collision_idx and x in out ):
            num_significant_digit += 1
            decimal.getcontext().prec = num_significant_digit
            x = str(decimal.Decimal( metaparameters[idx] ) + decimal.Decimal(0) )
        # if the last decimal digit of x ended with 0, we want make this digit be 1 i.e. 0.0920 -> 0.0921 
        out.append(x if x[-1] != '0' else x[:-1]+'1')
    return ','.join(out)



To demonstrate its output in some bad case,
if metaparameters = [0.23456,0.12345,0.23456,0.23457] (old collision exist)
Output (prompt user and leave collision as it is):
Some metaparameters may be same, which can be suspicous
'0.234,0.123,0.234,0.2345'

if metaparameters = [0.9999987,0.9999988,0.9999989,0.99999999,0.999972]
Output (no new collision):
‘0.999,0.9999,0.99999,0.999999,0.99997'

Similarly for case where there are 0s after decimal point,
metaparameters = [0.0000012344,0.00000123456,0.000001234567]
Output:
‘0.00000123,0.000001234,0.0000012345'

If metaparameters contains 0.0 or 1.0, it will notify user,
metaparameters = [0.0000012344,0.00000,1.000000]
Output:
Some metaparameters is 0.0 or 1.0, where is legitimate range is 0 < x < 1
‘0.00000123,0,1'

On Aug 17, 2016, at 7:54 AM, Wang Jian <notifications@github.com> wrote:

> In scripts/train_lm.py:
> 
> > -        if x == '0.00':
> > -            x = '{:.3g}'.format(param)
> > -        out.append(x)
> > -
> > +        temp_length = len(str(param))
> > +        if decimal_length < temp_length:
> > +            decimal_length = temp_length
> > +    decimal_length -= 2
> > +    
> > +    temp_original = metaparameters[0]
> > +    temp = round(temp_original, 3)
> > +    for param in metaparameters[1:]:
> > +        round_index = 3
> > +        x_original = param
> > +        x = round(x_original, round_index)
> > +        while(temp == x and round_index <= decimal_length):
> Hi Ke and Zhouyang, just a reminder in case you decide to try something similar with the code i posted earlier. What Dan said in this comment is right, we should use a while loop to avoid collision, since there could be a collision again after we append one digit. To avoid infinite loop, I think we can set a maximum value for num_non_zero_digit, like 7 or 8.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>